bigtable/client/table.cc

```diff
@@ -152,19 +161,84 @@ RowReader Table::ReadRows(RowSet row_set, std::int64_t rows_limit,
                        bigtable::internal::ReadRowsParserFactory>());
 }
 
-std::pair<bool, Row> Table::ReadRow(std::string row_key, Filter filter) {
+std::pair<bool, Row> Table::ReadRow(std::string row_key, Filter filter,
+                                    grpc::Status& ret_status) {
   RowSet row_set(std::move(row_key));
   std::int64_t const rows_limit = 1;
   RowReader reader =
-      ReadRows(std::move(row_set), rows_limit, std::move(filter));
+      ReadRows(std::move(row_set), rows_limit, std::move(filter), ret_status);
+  if (not ret_status.ok()) {
+    return std::make_pair(false, Row("", {}));
+  }
   auto it = reader.begin();
   if (it == reader.end()) {
+    grpc::Status status = reader.finish();
+    if (not status.ok()) {
+      ret_status = grpc::Status(status.error_code(), status.error_message());
+    }
     return std::make_pair(false, Row("", {}));
   }
   auto result = std::make_pair(true, std::move(*it));
   if (++it != reader.end()) {
-    internal::RaiseRuntimeError(
-        "internal error - RowReader returned 2 rows in ReadRow()");
+    ret_status =
+        grpc::Status(grpc::StatusCode::INTERNAL,
+                     "internal error - RowReader returned 2 rows in ReadRow()");
+    return std::make_pair(false, Row("", {}));
+  }
+  return result;
+}
+
+}  // namespace noex
+
+// Call the `google.bigtable.v2.Bigtable.MutateRow` RPC repeatedly until
+// successful, or until the policies in effect tell us to stop.
+void Table::Apply(SingleRowMutation&& mut) {
+  grpc::Status ret_status = grpc::Status::OK;
+  std::vector<FailedMutation> failures =
+      impl_.Apply(std::move(mut), ret_status);
+  if (not ret_status.ok()) {
+    ReportPermanentFailures(ret_status.error_message().c_str(), ret_status,
+                            failures);
+  }
+}
+// Call the `google.bigtable.v2.Bigtable.MutateRows` RPC repeatedly until
+// successful, or until the policies in effect tell us to stop.  When the RPC
+// is partially successful, this function retries only the mutations that did
+// not succeed.
+void Table::BulkApply(BulkMutation&& mut) {
+  grpc::Status ret_status = grpc::Status::OK;
+  std::vector<FailedMutation> failures =
+      impl_.BulkApply(std::move(mut), ret_status);
+  if (not ret_status.ok()) {
+    ReportPermanentFailures(ret_status.error_message().c_str(), ret_status,
+                            failures);
+  }
+}
+
+RowReader Table::ReadRows(RowSet row_set, Filter filter) {
+  grpc::Status ret_status = grpc::Status::OK;
+  auto result =
+      impl_.ReadRows(std::move(row_set), std::move(filter), ret_status);
+  return result;
+}
+
+RowReader Table::ReadRows(RowSet row_set, std::int64_t rows_limit,
+                          Filter filter) {
+  grpc::Status ret_status = grpc::Status::OK;
+  auto result = impl_.ReadRows(std::move(row_set), rows_limit,
+                               std::move(filter), ret_status);
+  if (not ret_status.ok()) {
+    internal::RaiseInvalidArgument(ret_status.error_message());
```

Comment:
While `RaiseInvalidArgument`?  That could be also `RaiseRuntimeError` depending on the type of failure, right?  We should consider removing the check on the `rows_limit` parameter, let the server tell us what is wrong with the parameters.

Reply:
Removed rows_limit check exception.